backend/src/schema/resolvers/searches.js

```diff
@@ -0,0 +1,74 @@
+import log from './helpers/databaseLogger'
+
+export default {
+  Query: {
+    findResources: async (_parent, args, context, _resolveInfo) => {
+      const { query, limit } = args
+      const { id: thisUserId } = context.user
+      // see http://lucene.apache.org/core/8_3_1/queryparser/org/apache/lucene/queryparser/classic/package-summary.html#package.description
+      const myQuery = query
+        .replace(/\s+/g, ' ')
+        .replace(/[[@#:*~\\$|^\]?/"'(){}+?!,.-;]/g, '')
+        .split(' ')
+        .map(s => (s.toLowerCase().match(/^(not|and|or)$/) ? '"' + s + '"' : s + '*'))
+        .join(' ')
+      const postCypher = `
+      CALL db.index.fulltext.queryNodes('post_fulltext_search', $query)
+      YIELD node as resource, score
+      MATCH (resource)<-[:WROTE]-(author:User)
+      WHERE score >= 0.5
+      AND NOT (
+        author.deleted = true OR author.disabled = true
+        OR resource.deleted = true OR resource.disabled = true
+        OR (:User { id: $thisUserId })-[:BLOCKED]-(author)
+      )
+      WITH resource, author,
+      [(resource)<-[:COMMENTS]-(comment:Comment) | comment] as comments,
+      [(resource)<-[:SHOUTED]-(user:User) | user] as shouter
+      RETURN resource {
+        .*,
+        __typename: labels(resource)[0],
+        author: properties(author),
+        commentsCount: toString(size(comments)),
+        shoutedCount: toString(size(shouter))
+      }
+      LIMIT $limit
+      `
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-12-20T16:01:46Z" title="Friday, December 20th 2019, 5:01:46 pm +01:00">Dec 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,69 @@
+const transformReturnType = record => {
+  return {
+    __typename: record.get('type'),
+    ...record.get('resource').properties,
+  }
+}
+export default {
+  Query: {
+    findResources: async (_parent, args, context, _resolveInfo) => {
+      const { query, limit } = args
+      const { id: thisUserId } = context.user
+      // see http://lucene.apache.org/core/8_3_1/queryparser/org/apache/lucene/queryparser/classic/package-summary.html#package.description
+      const myQuery = query + '*'
+      const postCypher = `
+      CALL db.index.fulltext.queryNodes('post_fulltext_search', $query)
+      YIELD node as resource, score
+      MATCH (resource)<-[:WROTE]-(user:User)
+      WHERE score >= 0.5
+      AND NOT (user.deleted = true OR user.disabled = true
+      OR resource.deleted = true OR resource.disabled = true
+      OR (:User { id: $thisUserId })-[:BLOCKED]-(user))
+      RETURN resource, labels(resource)[0] AS type
+      LIMIT $limit
+      `
+      const session = context.driver.session()
+      let postResults, userResults
+      const readPostTxResultPromise = session.readTransaction(async transaction => {
+        const postTransactionResponse = transaction.run(postCypher, {
+          query: myQuery,
+          limit,
+          thisUserId,
+        })
+        return postTransactionResponse
+      })
+      try {
+        postResults = await readPostTxResultPromise
```


We can improve performance here by waiting on both promises in parallel.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-12-20T16:13:48Z" title="Friday, December 20th 2019, 5:13:48 pm +01:00">Dec 20, 2019</time>_

---


I tried that and it threw an error saying that you could not start another transaction if there is already one pending...

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-12-20T19:15:35Z" title="Friday, December 20th 2019, 8:15:35 pm +01:00">Dec 20, 2019</time>_

---


Implemented now.